cdc/redo/meta_manager.go

```diff
@@ -0,0 +1,482 @@
+		changeFeedID:      contextutil.ChangefeedIDFromCtx(ctx),
+		uuidGenerator:     uuid.NewGenerator(),
+		enabled:           true,
+		flushIntervalInMs: cfg.FlushIntervalInMs,
+	}
+
+	uri, err := storage.ParseRawURL(cfg.Storage)
+	if err != nil {
+		return nil, err
+	}
+	if redo.IsBlackholeStorage(uri.Scheme) {
+		return m, nil
+	}
+
+	// "nfs" and "local" scheme is not supported by metaManager
+	redo.FixLocalScheme(uri)
+	extStorage, err := redo.InitExternalStorage(ctx, *uri)
+	if err != nil {
+		return nil, err
+	}
+	m.extStorage = extStorage
+	return m, nil
+}
+
+// Enabled returns whether this log manager is enabled
+func (m *metaManager) Enabled() bool {
+	return m.enabled
+}
+
+// Run runs bgFlushMeta and bgGC.
+func (m *metaManager) Run(ctx context.Context) error {
+	if m.extStorage == nil {
+		log.Warn("extStorage of redo meta manager is nil, skip running")
+		return nil
+	}
+
+	eg, egctx := errgroup.WithContext(ctx)
+	eg.Go(func() error {
+		return m.bgFlushMeta(egctx, m.flushIntervalInMs)
+	})
+	eg.Go(func() error {
+		return m.bgGC(egctx)
+	})
+	return eg.Wait()
+}
+
+// UpdateMeta updates meta.
+func (m *metaManager) UpdateMeta(checkpointTs, resolvedTs model.Ts) {
+	if ok := m.metaResolvedTs.checkAndSetUnflushed(resolvedTs); !ok {
+		log.Warn("update redo meta with a regressed resolved ts, ignore",
+			zap.Uint64("currResolvedTs", m.metaResolvedTs.getFlushed()),
+			zap.Uint64("recvResolvedTs", resolvedTs),
+			zap.String("namespace", m.changeFeedID.Namespace),
+			zap.String("changefeed", m.changeFeedID.ID))
+	}
+	if ok := m.metaCheckpointTs.checkAndSetUnflushed(checkpointTs); !ok {
+		log.Warn("update redo meta with a regressed checkpoint ts, ignore",
+			zap.Uint64("currCheckpointTs", m.metaCheckpointTs.getFlushed()),
+			zap.Uint64("recvCheckpointTs", checkpointTs),
+			zap.String("namespace", m.changeFeedID.Namespace),
+			zap.String("changefeed", m.changeFeedID.ID))
+	}
+}
+
+// GetFlushedMeta gets flushed meta.
+func (m *metaManager) GetFlushedMeta() common.LogMeta {
+	checkpointTs := m.metaCheckpointTs.getFlushed()
+	resolvedTs := m.metaResolvedTs.getFlushed()
+	return common.LogMeta{CheckpointTs: checkpointTs, ResolvedTs: resolvedTs}
+}
+
+// initMeta will read the meta file from external storage and initialize the meta
+// field of metaManager.
+func (m *metaManager) initMeta(ctx context.Context, startTs model.Ts) error {
+	select {
+	case <-ctx.Done():
+		return errors.Trace(ctx.Err())
+	default:
+	}
+
+	metas := []*common.LogMeta{
+		{CheckpointTs: startTs, ResolvedTs: startTs},
+	}
+	var toRemoveMetaFiles []string
+	err := m.extStorage.WalkDir(ctx, nil, func(path string, size int64) error {
+		// TODO: use prefix to accelerate traverse operation
+		if !strings.HasSuffix(path, redo.MetaEXT) {
+			return nil
+		}
+		toRemoveMetaFiles = append(toRemoveMetaFiles, path)
+
+		data, err := m.extStorage.ReadFile(ctx, path)
+		if err != nil && !util.IsNotExistInExtStorage(err) {
```

Comment:
Maybe we can use the `FileExists()` of `ExternalStorage` interface to simplify this check?

Reply:
The check operation here must be atomic.